### PR TITLE
postgres v3 experimental

### PR DESCRIPTION
**Pull request by isoos**
This is a draft version of how the v3 of the API could look like. Some highlights and changes over the current design:
- Integrating `package:postgres_pool` into the core API.
- Only a single prepare -> start (streaming the results) -> wait for completion mechanism for querying (with helper `execute` method for the simple use case to get all of the rows).
- I'm unsure about `PgResultSchema` (how efficient we can make the iod -> name resolution, esp. for table names). We may need to have an extra parameter that could skip the resolution if it is not needed.
- explicit `PgSession` abstraction that could be used inside a transaction/retry callback or even top-level
- pool could also expose the raw connection object

### CODE REVIEW

File: lib/postgres_v3_experimental.dart [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+  Future<PgStatement> prepare(
+    String sql, {
+    Object? /* String */ substitution,
+    Object? /* List<PgDataType> | Map<String, PgDataType> */ types,
+    Duration? timeout,
+  });
+
+  Future<PgResult> execute(
+    String sql, {
+    Object? /* String */ substitution,
```
Comment (simolus3) [inline lib/postgres_v3_experimental.dart:31]: What does `substitution` do? I guess it's related to `types` and `parameters`, but in which way exactly? Don't these two provide all the information we need?

File: lib/postgres_v3_experimental.dart [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+  PgMessages get messages;
+}
+
+abstract class PgResultStream implements Stream<PgResultRow> {
+  Future<int> get affectedRows;
+  Future<PgResultSchema> get schema;
+}
+
+abstract class PgStatement {
+  PgResultStream start(
```
Comment (simolus3) [inline lib/postgres_v3_experimental.dart:76]: I like the idea of being able to run queries without caching all results in memory. But we also can't implement pauses/cancellations efficiently in this protocol, so I'm not sure if `Stream` is the right interface to use? Maybe this should be a callback, even if that generally feels less Dart-like? I'm not sure about this though. Perhaps it is enough to document that pauses are emulated for technical reasons.

**Review by simolus3 — COMMENTED** [review]
I really like this API!

File: lib/postgres_v3_experimental.dart (hunk 1 above)
Comment (isoos) [reply lib/postgres_v3_experimental.dart:31]: I thought that to get flexibility in the query substitution, we could have both `?` or `:name` or `@name` as substitution pattern. e.g. `substitution: ':'` would enable `:name` replacement patterns, keeping the `@` out of harms way...

File: lib/postgres_v3_experimental.dart (hunk 2 above)
Comment (isoos) [reply lib/postgres_v3_experimental.dart:76]: Not efficient, but first version could be just ignoring the rest of the results after cancelling the stream, but later on we could do something like this:
https://stackoverflow.com/questions/35319597/how-to-stop-kill-a-query-in-postgresql

Callbacks could work too, and potentially simpler to implement.

Comment (simolus3) [reply lib/postgres_v3_experimental.dart:76]: If there's an eventual goal to support that I think streams are a good API. I was just concerned whether we could pull this off at all.

File: lib/postgres_v3_experimental.dart (hunk 1 above)
Comment (simolus3) [reply lib/postgres_v3_experimental.dart:31]: I think it's a good idea to allow multiple formats, but at least to me it was confusing that they are all parameters on the same method, especially since they appear to be somewhat incompatible with each other. Maybe we could have something like

```dart
class PgQueryDescription {
  factory PgQueryDescription.direct(String sql, {List<PgDataType>? types}); // direct postgres api
  factory PgQueryDescription.map(String sql, {String substitution = '@'}); // local post-processing
}
```

And then `prepare` would take a `PGQueryDescription` instead of `sql` + `types` + `substitution`? To make the default case easier, we could also have it take an `Object /* String | PgQueryDescription */`, defaulting to `PgQueryDescription.map` if given a string.

**Comment by simolus3** [comment]
Are there plans to implement this API already? I'd love to help out and get started otherwise.

File: lib/postgres_v3_experimental.dart (hunk 2 above)
Comment (isoos) [reply lib/postgres_v3_experimental.dart:76]: I'm tempted to backpedal on the `Stream` design, and rather use a plain callback-based method, partly because it seems to me that the above-mentioned query killer is really a backend/connection killer.

File: lib/postgres_v3_experimental.dart (hunk 1 above)
Comment (isoos) [reply lib/postgres_v3_experimental.dart:31]: Good points here. I tend to agree (as the alternative would be `prepareDirect`, `prepareMap`, ...), although I'd probably go with a much shorter name, like `PgSql`, as this would be used frequently.

Another angle for this would be to configure the substitution only at the connection/session level.

**Comment by isoos** [comment]
> Are there plans to implement this API already? I'd love to help out and get started otherwise.

This is a best effort on my side, any help is appreciated. However, I think if this gets done, it should (a) provide backward API compatibility for clients to allow gradual migration, and (b) be solid enough that we won't need another API redesign in the next few years. So I don't mind taking some time contemplating on the API, but if there is a consensus in most of the things, I won't put further barriers into getting things done/merged.

File: lib/postgres_v3_experimental.dart (hunk 2 above)
Comment (simolus3) [reply lib/postgres_v3_experimental.dart:76]: We can send an execute message with a fixed maximum row count (say 25). If there are more rows, postgres will send a `PortalSuspended` message informing us that more rows are available once we send another execute message.

So I think we could implement cancellations by just closing the portal and pauses by holding off sending execute messages.

----

Of course, we can also just add the non-streaming API first and then add a stream method once we have figured out how to implement it.

Comment (isoos) [reply lib/postgres_v3_experimental.dart:76]: Apparently I'm unaware of many details of the postgres protocols, it seems to be a great suggestion! For `Stream` we could use a `StreamController` internally and its `onPause`, `onResume` methods to control the flow. I think it wouldn't be much different from the callbacks.

**Comment by simolus3** [comment]
@isoos Do you think it would be possible to release a version with the v3 library draft to pub.dev already? I know that not all v3 APIs are implemented yet, but it's hidden behind an experimental import so it might be fine? Even with no stability guarantees, having this on pub.dev would be great to publish some experimental wrappers around the new API.

If it helps, I'm happy to open a PR bumping the version and adding a changelog. Is there anything else I can do to help getting this published?

**Comment by isoos** [comment]
> Do you think it would be possible to release a version with the v3 library draft to pub.dev already?

Hm, I don't mind that, but I'm unsure what version it should be... `2.7.0-dev.1` maybe? It is not a prerelease for 3.0.0 yet, as the main API is still on v2...

**Comment by eseidel** [comment]
I would be happy to help, but it's unclear how to contribute to this effort.  Otherwise I'll just keep sending you small fixes to v2 for now. 🤷

File: lib/postgres_v3_experimental.dart [hunk 3]
```diff
@@ -0,0 +1,204 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+
+abstract class PgPool implements PgSession, PgSessionExecutor {
```
Comment (eseidel) [inline lib/postgres_v3_experimental.dart:6]: It's unfortunate that Dart pushes us towards namespacing class names (like Obj-C did).  One could go the http route, and just use the good names and expect everyone to import you via import 'package:postgres/postgres.dart' as pg;

File: lib/postgres_v3_experimental.dart [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  static Future<PgPool> open(
+    List<PgEndpoint> endpoints, {
+    PgSessionSettings? sessionSettings,
+    PgPoolSettings? poolSettings,
+  }) =>
+      throw UnimplementedError();
+
+  Future<R> withConnection<R>(
```
Comment (eseidel) [inline lib/postgres_v3_experimental.dart:14]: This promotes a single connection into being a pool? (e.g. the first connection of several)?

File: lib/postgres_v3_experimental.dart (hunk 3 above)
Comment (isoos) [reply lib/postgres_v3_experimental.dart:6]: I like the `package:http`-way too, but in the end it doesn't give much difference to have `pg.X` or `PgX`.

Comment (eseidel) [reply lib/postgres_v3_experimental.dart:6]: Well, the big difference is in the optionality.  Inside package:http they don't have to use http.  And for places where including package:http doesn't conflict with "Client" or "Request" or "get", etc people happily can.

So if you like the package http way, I'd encourage you to consider it here.  You'd delete a ton of code from inside the package. 🤷

Comment (eseidel) [reply lib/postgres_v3_experimental.dart:6]: Rust has gotten me on this pattern.  Rust has strong (and nicer than c++) namespaces and encourages you to just always use nice names for everything.  Obj-C of course (and I guess Dart, sorta?) doesn't, just has a flat namespace, so you were required to come up with your own globally-safe prefix for everything.

**Comment by osaxma** [comment]
Hello Everyone,

Sorry if I am late for the game as I just had a chance to review the new changes. 

I would like to comment on the Logical Replication Protocol support in v3. I noticed that the APIs that allowed utilizing the replication protocol are not public anymore -- namely the APIs for accessing the socket connection to send and receive server messages. I would like to give some background information as a refresher and also mention what is needed to continue supporting the replication protocol in v3. 

When I added the implementation for the Logical Replication Protocol (purely as a method to obtain realtime functionality by listening to changes in database tables -- similar to how Supabase and others do it), I did not add a friendly high level public API. So, as a user, I relied on the exposed `PostgreSQLConnection.addMessage` and `PostgreSQLConnection.stream` of v2 to communicate with the server externally as shown in this [example](https://github.com/osaxma/postgresql-dart-replication-example/blob/main/example/listen.dart). 

I also like mention two other points about how the logical replication protocol was used in v2 to provide a full picture. The first point is that in order to activate the Logical Replication Protocol, the user must pass a flag when establishing the connection which is then passed to the `StartupMessage`. Currently the flag is not implemented in v3 but I saw that it's marked as a [todo item](https://github.com/isoos/postgresql-dart/blob/81acc742f1219e151e2778fd31530143cf43b1c2/lib/src/v3/connection.dart#L307) in `PgConnectionImplementation._startup()` within `v3/connection.dart`. The second point is that when the user listens to server messages in logical replication mode, they only handle messages related to logical replications (i.e. StartUp and Close messages are handled by the package itself). 


To summarize the previous points, three features are needed in order to continue supporting the Logical Replication Protocol in v3:
- Ability to pass a flag to activate logical replication
- Ability to listen to server messages
- Ability to send messages to the server 

Though I am not sure what is the preferable approach to implement them. The first one is obvious but the other two requires an API design decision (e.g. re-expose stream/addMessage or have the user pass a message handler object). 

Let me know what do you think and how can I help.

**Comment by isoos** [comment]
> Ability to pass a flag to activate logical replication
> Ability to listen to server messages
> Ability to send messages to the server

@osaxma: I'm wondering if these should live on a separate kind of connection-like class, using the same transport layer, but exposing different kind of API? Because the generic connection interface may be used in a pool context, where the exact server you send the query to may not matter, while here both the event listener and the event sender is targeting a single server.

**Comment by osaxma** [comment]
> @osaxma: I'm wondering if these should live on a separate kind of connection-like class, using the same transport layer, but exposing different kind of API? Because the generic connection interface may be used in a pool context, where the exact server you send the query to may not matter, while here both the event listener and the event sender is targeting a single server.

@isoos That's a good point I haven't considered. I will try to put something together soon and share it as a draft PR for feedback.

**Comment by simolus3** [comment]
@osaxma Note that you can supply a `StreamChannelTransformer` for the connection, which can be used to inject and listen to packets.

Here is an example that just prints every packet, but you can write your own sink transformer that captures the sink to add your own messages as well as setting the replication flags by intercepting startup packets. Similarly, you can use the stream transformers to listen to incoming replication messages without forwarding them to the v3 client implementation.

https://github.com/isoos/postgresql-dart/blob/9d5fcd4792f64434c41f02399f91362505578043/test/v3_test.dart#L26-L44
